compiler/noirc_frontend/src/elaborator/scope.rs

```diff
@@ -43,7 +43,34 @@ impl<'context> Elaborator<'context> {
     }
 
     pub(super) fn resolve_path(&mut self, path: Path) -> Result<ModuleDefId, ResolverError> {
-        let resolver = StandardPathResolver::new(self.module_id());
+        let mut module_id = self.module_id();
+        let mut path = path;
+
+        if path.kind == PathKind::Plain && path.first_name() == SELF_TYPE_NAME {
+            if let Some(Type::Struct(struct_type, _)) = &self.self_type {
```

Comment:
Should we output a specific user error if `Self` points to an unresolvable `Type`?

Reply:
Right now the error is "Could not resolve 'Self' in path". 

Checking Rust, if you have this:

```rust
fn foo() -> Self {}
```

the error is "cannot find type `Self` in this scope" with some more helpful comments like "`Self` is only available in impls, traits, and type definitions" and "`Self` not allowed in a function".

If you have this:

```rust
fn foo() {
    Self::bar();
}
```

the error is "failed to resolve: `Self` is only available in impls, traits, and type definitions".

I guess in both cases we could say "cannot find type `Self` in this scope. `Self` is only available in impls, traits, and type definitions".

What do you think?

Reply:
Put another way, do you think a more specific error than "Could not resolve 'Self' in path" is worth it? The current error message isn't wrong either...

Reply:
Rust's error is better and if possible we should emulate it. That being said, I think our error is also fine and we don't need to bend over backwards trying to emulate rust's error if it is overly cumbersome to do so.